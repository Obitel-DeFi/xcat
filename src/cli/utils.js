import chalk from 'chalk'
import {existsSync} from 'fs'

import Trade from '../trade'
import Config from '../config'
import {fileToObj} from '../utils'

const logError = msg => console.error(chalk.red(`\nERROR: ${msg}`))

const verifyArgTradeFile = filename => {
  if (!filename || (typeof filename === 'string' && filename.trim() === '')) {
    logError(`No file provided. Pass the name of a trade.json file.`)
    return false
  }

  if (!existsSync(filename)) {
    logError(`Trade file [${filename}] does not exist.`)
    return false
  }

  try {
    new Trade(fileToObj(filename))
  } catch (err) {
    logError(`Failed to parse [${filename}]:`)
    console.error(`\n\n${err}`)
    return false
  }

  return true
}

const verifyConfigFile = filename => {
  // look for config.json in local directory if not provided
  if (!filename || typeof filename !== 'string' || filename.trim() === '')
    filename = './config.json'

  if (!existsSync(filename)) {
    logError(
      `Config file [${filename}] does not exist. Put a config.json in the ` +
        `local path OR provide one with --config.`
    )
    return false
  }

  try {
    new Config(fileToObj(filename))
  } catch (err) {
    logError(`Failed to parse [${filename}]:`)
    console.error(`\n\n${err}`)
    return false
  }

  return true
}

const verifyArgTradeID = id => {
  throw new Error(
    `not yet implemented - will pull a trade record for ${id} from a local database`
  )
}

export {
  fileToObj,
  logError,
  verifyArgTradeFile,
  verifyArgTradeID,
  verifyConfigFile,
}
