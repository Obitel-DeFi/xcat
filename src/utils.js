import crypto from 'crypto'
import {URL} from 'url'
import stellarSdk from 'stellar-sdk'
import Web3Utils from 'web3-utils'
import hasIn from 'lodash/hasIn'

// Format required for sending bytes through an eth js client:
//  - hex string representation
//  - prefixed with 0x
const bufToStr = b => '0x' + b.toString('hex')

const clone = obj => JSON.parse(JSON.stringify(obj))

const sha256 = x =>
  crypto
    .createHash('sha256')
    .update(x)
    .digest()

const random32 = () => crypto.randomBytes(32)

const newSecretHashPair = () => {
  const secret = random32()
  const hash = sha256(secret)
  return {
    secret: bufToStr(secret),
    hash: bufToStr(hash),
  }
}

const isStellarPublicAddress = addr =>
  stellarSdk.StrKey.isValidEd25519PublicKey(addr)
const isStellarSecretSeed = addr =>
  stellarSdk.StrKey.isValidEd25519SecretSeed(addr)
const isStellarFederatedAddress = addr => /^[^*,]*\*[a-z0-9-.]*$/i.test(addr)

const isEthereumPublicAddress = addr => Web3Utils.isAddress(addr)

const isUrl = url => {
  try {
    new URL(url)
    return true
  } catch (e) {
    return false
  }
}

const isClassWithName = (cls, name) =>
  hasIn(cls, 'constructor.name') && cls.constructor.name === name

export {
  bufToStr,
  clone,
  newSecretHashPair,
  random32,
  sha256,
  isStellarPublicAddress,
  isStellarSecretSeed,
  isStellarFederatedAddress,
  isEthereumPublicAddress,
  isClassWithName,
  isUrl,
}
