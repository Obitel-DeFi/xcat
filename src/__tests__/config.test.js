import expect from 'expect'
import {clone} from '../utils'
import Config from '../config'

const config1 = require('./__data__/config1.json')

describe('config', () => {
  it('creates new Config that has all properties of the given config object', () => {
    const config = new Config(config1)
    Object.keys(config1).forEach(attr =>
      expect(config[attr]).toEqual(config1[attr])
    )
  })

  it('throws if config has a bad stellarPublicAddress', () => {
    const configBad = clone(config1)
    configBad.stellarAccountSecret = 'SCABCDEF'
    expect(() => new Config(configBad)).toThrowErrorMatchingSnapshot()
  })

  it('throws if config has a bad ethereumPublicAddress', () => {
    const configBad = clone(config1)
    configBad.ethereumPublicAddress = '0xabcdef'
    expect(() => new Config(configBad)).toThrowErrorMatchingSnapshot()
  })

  it('throws if unknown stellarNetwork', () => {
    const configBad = clone(config1)
    configBad.stellarNetwork = 'private'
    expect(() => new Config(configBad)).toThrowErrorMatchingSnapshot()
  })

  it('throws if config has a bad ethereumRPC', () => {
    const configBad = clone(config1)
    configBad.ethereumRPC = '//notaurl'
    expect(() => new Config(configBad)).toThrowErrorMatchingSnapshot()
  })

  it('provides property reference to derived stellar public key', () => {
    const configJSON = clone(config1)
    const config = new Config(configJSON)
    expect(config.stellarPublicAddress).toEqual(
      'GCOCD2SVBEOB4PWMS7FEOFM27TPSZBO5LFR24NBMT6Q73OIUKBKDKFBS'
    )
  })
})
