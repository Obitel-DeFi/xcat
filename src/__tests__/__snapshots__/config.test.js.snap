// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config throws if config has a bad ethereumPublicAddress 1`] = `
"config doesn't conform to schema config.json [
	message: \\"data.ethereumPublicAddress should match format \\"ethereumPublicAddress\\"\\"
	raw: [
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".ethereumPublicAddress\\",
    \\"schemaPath\\": \\"#/properties/ethereumPublicAddress/format\\",
    \\"params\\": {
      \\"format\\": \\"ethereumPublicAddress\\"
    },
    \\"message\\": \\"should match format \\\\\\"ethereumPublicAddress\\\\\\"\\"
  }
]]"
`;

exports[`config throws if config has a bad ethereumRPC 1`] = `
"config doesn't conform to schema config.json [
	message: \\"data.ethereumRPC should match format \\"URL\\"\\"
	raw: [
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".ethereumRPC\\",
    \\"schemaPath\\": \\"#/properties/ethereumRPC/format\\",
    \\"params\\": {
      \\"format\\": \\"URL\\"
    },
    \\"message\\": \\"should match format \\\\\\"URL\\\\\\"\\"
  }
]]"
`;

exports[`config throws if config has a bad stellarPublicAddress 1`] = `
"config doesn't conform to schema config.json [
	message: \\"data.stellarAccountSecret should match format \\"stellarSecretSeed\\"\\"
	raw: [
  {
    \\"keyword\\": \\"format\\",
    \\"dataPath\\": \\".stellarAccountSecret\\",
    \\"schemaPath\\": \\"#/properties/stellarAccountSecret/format\\",
    \\"params\\": {
      \\"format\\": \\"stellarSecretSeed\\"
    },
    \\"message\\": \\"should match format \\\\\\"stellarSecretSeed\\\\\\"\\"
  }
]]"
`;

exports[`config throws if unknown stellarNetwork 1`] = `
"config doesn't conform to schema config.json [
	message: \\"data.stellarNetwork should be equal to one of the allowed values\\"
	raw: [
  {
    \\"keyword\\": \\"enum\\",
    \\"dataPath\\": \\".stellarNetwork\\",
    \\"schemaPath\\": \\"#/properties/stellarNetwork/enum\\",
    \\"params\\": {
      \\"allowedValues\\": [
        \\"public\\",
        \\"testnet\\",
        \\"localnet\\"
      ]
    },
    \\"message\\": \\"should be equal to one of the allowed values\\"
  }
]]"
`;
